runtime-light/stdlib/component/inter-component-session/client.h

```diff
@@ -0,0 +1,291 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2025 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <expected>
+#include <iterator>
+#include <memory>
+#include <optional>
+#include <span>
+#include <string_view>
+#include <utility>
+
+#include "runtime-light/coroutine/event.h"
+#include "runtime-light/coroutine/task.h"
+#include "runtime-light/stdlib/component/inter-component-session/buffer-provider.h"
+#include "runtime-light/stdlib/diagnostics/logs.h"
+#include "runtime-light/streams/stream.h"
+#include "runtime-light/tl/tl-core.h"
+#include "runtime-light/tl/tl-functions.h"
+#include "runtime-light/tl/tl-types.h"
+
+namespace kphp::component::inter_component_session {
+
+// Inter component client for interactions over a stream in a client-server manner
+class Client final {
+  using query_id_type = uint64_t;
+  using query2notifier_type = kphp::stl::unordered_map<query_id_type, kphp::coro::event, kphp::memory::script_allocator>;
+
+  struct refcountable_transport_type : public refcountable_php_classes<refcountable_transport_type> {
+    kphp::component::stream stream;
+  };
+  using shared_transport_type = class_instance<refcountable_transport_type>;
+
+  shared_transport_type transport;
+  query_id_type query_count{};
+
+  // WRITER
+  struct writer {
+    bool is_occupied{false};
+    query_id_type occupied_by{0};
+    query2notifier_type query2notifier{};
+    kphp::stl::queue<query_id_type, kphp::memory::script_allocator> queue;
+
+    inline auto write(shared_transport_type t, query_id_type qid,
+                      std::span<const std::byte> payload) noexcept -> kphp::coro::task<std::expected<void, int32_t>> {
```

Comment:
Maybe it's better to just return int32_t? In any case, the caller should know the semantics of the error code, so then you can use the fact that k2::errno_ok is the success code.